File: cpp/tennis1.cc
#include <string>
#include <map>

const std::string getIndividualScore(int score)
{
    constexpr const char* scoreList[4] = {"Love", "Fifteen", "Thirty", "Forty"};
    return scoreList[score];
}

const std::string getWinOrAdvantage(int scoreDifference)
{
    std::map<int, std::string> outcomes = {
        {1, "Advantage player1"}, {-1, "Advantage player2"}, {2, "Win for player1"}, {-2, "Win for player2"}};

    auto it = outcomes.find(scoreDifference);
    if (it != outcomes.end())
    {
        return it->second;
    } 
    else
    {
        return scoreDifference > 0 ? outcomes[2] : outcomes[-2];
    }
}

inline const std::string getEqualScore(int p1CurrentScore, int deuceCriteria = 3)
{
    return (p1CurrentScore >= deuceCriteria) ? "Deuce" : getIndividualScore(p1CurrentScore) + "-All";
}

inline const std::string getCurrentScoreBetweenPlayers(int p1CurrentScore, int p2CurrentScore)
{
    return getIndividualScore(p1CurrentScore) + "-" + getIndividualScore(p2CurrentScore);
}

inline const int isMatchPoint(int p1CurrentScore, int p2CurrentScore, int matchPointCriteria = 4)
{
    return p1CurrentScore >= matchPointCriteria || p2CurrentScore >= matchPointCriteria;
}

inline const int calculateScoreDifference(int p1CurrentScore, int p2CurrentScore)
{
    return p1CurrentScore - p2CurrentScore;
}

const std::string tennisScore(int p1CurrentScore, int p2CurrentScore)
{
    bool isGameTied = p1CurrentScore == p2CurrentScore ? true : false;
    if (isGameTied)
    {
        return getEqualScore(p1CurrentScore);
    }
    else if (isMatchPoint(p1CurrentScore, p2CurrentScore))
    {
        return getWinOrAdvantage(calculateScoreDifference(p1CurrentScore, p2CurrentScore));
    }
    else
    {
        return getCurrentScoreBetweenPlayers(p1CurrentScore, p2CurrentScore);
    }
}

File: python/tennis1.py
from enum import IntEnum


class Points(IntEnum):
    LOVE = 0
    FIFTEEN = 1
    THIRTY = 2
    FORTY = 3

    def pretty(self):
        return self.name.title()


class TennisGame1:
    def __init__(self, player1_name, player2_name):
        self.player1_name = player1_name
        self.player2_name = player2_name
        self.p1_points = Points.LOVE
        self.p2_points = Points.LOVE

    def won_point_player1(self):
        self.p1_points += 1

    def won_point_player2(self):
        self.p2_points += 1

    def score(self):
        result = ""
        match (self.p1_points, self.p2_points):
            case [p1, p2] if p1 == p2 and p1 < Points.FORTY:
                # Even points, below forty
                result = f"{Points(p1).pretty()}-All"
            case [p1, p2] if p1 == p2:
                # Even points, forty or above
                result = "Deuce"
            case [p1, p2] if p1 <= Points.FORTY and p2 <= Points.FORTY:
                # Different points, both below forty
                result = f"{Points(p1).pretty()}-{Points(p2).pretty()}"
            case [p1, p2] if p1 - p2 == 1:
                # One ball difference, above forty, player 1 advantage
                result = f"Advantage {self.player1_name}"
            case [p1, p2] if p2 - p1 == 1:
                # One ball difference, above forty, player 2 advantage
                result = f"Advantage {self.player2_name}"
            case [p1, p2] if p1 - p2 > 1:
                # More than one ball difference, player 1 above forty
                result = f"Win for {self.player1_name}"
            case [p1, p2] if p2 - p1 > 1:
                # More than one ball difference, player 2 above forty
                result = f"Win for {self.player2_name}"
        return result
